README.md
# Vercel-Serverless-Api

## Code Quality Status
![Build Status](https://github.com/gastonpereyra/vercel-serverless-api/workflows/Build%20Status/badge.svg)
[![Coverage Status](https://img.shields.io/coveralls/github/gastonpereyra/vercel-serverless-api/master.svg)](https://coveralls.io/r/gastonpereyra/vercel-serverless-api?branch=master)

![npm-vercel-serverless-api](https://user-images.githubusercontent.com/39351850/132961710-27605cee-2e42-410a-bc08-cbf091094873.png)

## Description
A handler for Serverless Function in Vercel to develop API

## Installation

```
npm i vercel-serverless-api
```

## API

Its a Class to help to create an API.

### Getters

#### Ids
* `pathIds`: _object_, Query Parameters or Path-Parameters for ids
    * Example: `https://example.verce.app/api/message?pathIds.id=10`
        * `pathIds.id`: '10'
    * Example: `https://example.verce.app/api/message/?pathIds.emailId=11`
        * `pathIds.emailId`: '11'

#### Body
* `data`: _object_, Body

#### Queries
* `filters`: _object_, Query parameters to filter
    * Example `https://example.verce.app/api/message?name=John&filters.age=10`
        * `filters.name`: 'John'
        * `filters.age`: '10'
* `this.sort`: _object_, Query parameters to sort
    * Example: `https://example.verce.app/api/message?sortBy=name&sortDirection=desc`
        * `sort.by`: 'name'
        * `sort.direction`: 'desc'

* `query`: _object_, Query parameters
    * Example: `https://example.verce.app/api/message?other.foo=name`
        * `query.foo`: 'name'

#### Request

Other request data

* `request`
    * `request.url`: Request URL
    * `request.method`: Request REST Method
    * `request.headers`: Request Headers
    * `request.cookies`: Request cookies

### Methods

* `setCode(code)`: To setup a custom response status-code
    * `code`: _number_

* `setHeader(header, value)`: To setup a custom response header.
    * `header`: _string_
    * `value`: _string_ or _number_ or _boolean_

* `setBody(body)`: To setup a custom response body. If you do not set a custom Content-type, this will be `application/json`
    * `body`: _object_ (for JSON) or _any_ (for other content-type)

* `validate`: For validation. If you throw an error, will setup status-code 400 by default
    * `async`

* `process`: The API itself. If you throw an error, will setup status-code 500 by default
    * `async`

### Structure Validation

Can use [`superstruct@0.7.0`](https://github.com/ianstormtaylor/superstruct/tree/v0.7.0) to validate body, ids, filters, sort, only must rewrite the following method

* `idsStruct`
    * `static`
* `bodyStruct`
    * `static`
* `filtersStruct`
    * `static`
* `sortStruct`
    * `static`
* `queryStruct`
    * `static`

### Usage

```js
const { struct } = require('superstruct'); // only works up to 0.7.0 version
const { API } = require('vercel-serverless-api');

module.exports = class MyApi extends API {

    static get idsStruct() {
        return struct({
            id: 'string'
        });
    }

    static get bodyStruct() {
        return struct({
            name: 'string',
            age: 'string?'
        });
    }

    static get filtersStruct() {
        return struct({
            name: 'string|null?',
            age: 'string|null?'
        });
    }

    static get sortStruct() {
        return struct({
            by: 'string?',
            direction: 'string?'
        });
    }

    validate() {

        if(this.data.age < 10)
            throw new Error('Too Young'); // statusCode will be 400
    }

    process() {

        if(!this.data.name)
            throw new Error('Empty String is not valid'); // statusCode will be 500

        if(this.data.image) {
            this.setHeader('Content-Type', 'text/plain') // Set a Custom Content-Type
                .setHeader('X-Custom', 100) // Set a Custom Header
                .setBody('<h1>Secret</h1>') // Because setting a Custom Type will Response a Plain Text
        }

        this.setCode(201).setBody({
            name: this.data.name
            lastname: 'Stark',
            age: this.data.age + 1
        });
    }
}

```

## Handler

The API Class and Handler can be combined to help to devolope Serverless Function in Vercel

```js
// in ./api/message/post.js
const { hanlder } = require('vercel-serverless-api');

const MyApi = require('./my-api');

module.exports = (..args) => handler(MyApi, ...args);
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2021-09-18
## Added
- Support for custom response-headers and content-type

## [1.0.0] 2021-09-12
### Added
- Serverless Function Handler
- API Class


tests/handler.js
'use strict';

const assert = require('assert');
const sinon = require('sinon');
const { struct } = require('@janiscommerce/superstruct');

const { handler, API } = require('../lib/index');

describe('Vercel Serverless API', () => {

	const responseDummy = {
		status() { return this; },
		json() { return this; },
		send() { return this; },
		setHeader() { return this; }
	};

	beforeEach(() => {
		sinon.spy(responseDummy, 'status');
		sinon.spy(responseDummy, 'json');
		sinon.spy(responseDummy, 'send');
		sinon.spy(responseDummy, 'setHeader');
	});

	afterEach(() => {
		sinon.restore();
	});

	context('When cannot use the handler', () => {

		it('Should rejects if response is not passed', async () => {
			await assert.rejects(() => handler(null, {}));
		});

		it('Should returns 500 and Error Message if API Class is not passed', async () => {
			await handler(null, {}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 500);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Internal Server Error',
				error: 'API cannot be initialize. Cannot found API Class'
			});
		});
	});

	context('When handler can validate structure', () => {

		it('Should return 200 and empty body if API can passed ids struct', async () => {

			class ApiDummy extends API {

				static get idsStruct() {
					return struct.partial({
						id: 'number'
					});
				}
			}

			await handler(ApiDummy, {
				query: {
					'pathIds.id': 100
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 200);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {});
		});

		it('Should return 400 and error if API cannot passed ids struct', async () => {

			class ApiDummy extends API {

				static get idsStruct() {
					return struct.partial({
						id: 'number'
					});
				}
			}

			await handler(ApiDummy, {
				query: {
					'pathIds.id': 'a-100'
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 400);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Invalid Struct for Ids',
				error: [{
					field: 'id',
					type: 'number',
					value: 'a-100'
				}]
			});
		});

		it('Should return 200 and empty body if API can passed body struct', async () => {

			class ApiDummy extends API {

				static get bodyStruct() {
					return struct.partial({
						name: 'string'
					});
				}
			}

			await handler(ApiDummy, {
				body: {
					name: 'Jonh'
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 200);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {});
		});

		it('Should return 400 and error if API cannot passed body struct', async () => {

			class ApiDummy extends API {

				static get bodyStruct() {
					return struct.partial({
						name: 'string'
					});
				}
			}

			await handler(ApiDummy, {
				body: {
					name: 100
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 400);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Invalid Struct for Body',
				error: [{
					field: 'name',
					type: 'string',
					value: 100
				}]
			});
		});

		it('Should return 200 and empty body if API can passed filters struct', async () => {

			class ApiDummy extends API {

				static get filtersStruct() {
					return struct.partial({
						age: 'number',
						name: 'string'
					});
				}
			}

			await handler(ApiDummy, {
				query: {
					'filters.age': 100,
					name: 'Johnd'
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 200);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {});
		});

		it('Should return 400 and error if API cannot passed filters struct', async () => {

			class ApiDummy extends API {

				static get filtersStruct() {
					return struct.partial({
						age: 'number',
						name: 'string'
					});
				}
			}

			await handler(ApiDummy, {
				query: {
					'filters.age': 'Hi',
					name: 10
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 400);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Invalid Struct for Filters',
				error: [
					{
						field: 'age',
						type: 'number',
						value: 'Hi'
					},
					{
						field: 'name',
						type: 'string',
						value: 10
					}
				]
			});
		});

		it('Should return 200 and empty body if API can passed sort struct', async () => {

			class ApiDummy extends API {

				static get sortStruct() {
					return struct.partial({
						by: 'string',
						direction: 'string'
					});
				}
			}

			await handler(ApiDummy, {
				query: {
					sortBy: 'name',
					sortDirection: 1
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 200);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {});
		});

		it('Should return 400 and error if API cannot passed sort struct', async () => {

			class ApiDummy extends API {

				static get sortStruct() {
					return struct.partial({
						by: 'string',
						direction: 'string'
					});
				}
			}

			await handler(ApiDummy, {
				query: {
					sortBy: 1,
					sortDirection: 1
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 400);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Invalid Struct for Sort',
				error: [{
					field: 'by',
					type: 'string',
					value: 1
				}]
			});
		});

		it('Should return 200 and empty body if API can passed query struct', async () => {

			class ApiDummy extends API {

				static get queryStruct() {
					return struct.partial({
						content: {
							total: 'number'
						}
					});
				}
			}

			await handler(ApiDummy, {
				query: {
					'content.total': 10
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 200);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {});
		});

		it('Should return 400 and error if API cannot passed query struct', async () => {

			class ApiDummy extends API {

				static get queryStruct() {
					return struct.partial({
						content: {
							total: 'number'
						}
					});
				}
			}

			await handler(ApiDummy, {
				query: {
					'content.total': 'Ten'
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 400);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Invalid Struct for Query',
				error: [{
					field: 'content.total',
					type: 'number',
					value: 'Ten'
				}]
			});
		});
	});

	context('When handler can execute validate from API', () => {

		class ApiDummy extends API {

			validate() {
				if(this.data.name === 'John')
					throw new Error('Wrong name!');

				if(this.data.age === 10) {
					const error = new Error('Wrong age');
					error.extraMessage = 'Not 10';

					throw error;
				}

				if(this.data.content.houseId === 10) {
					this.setCode(404);
					throw new Error('Not found');
				}
			}
		}

		it('Should return 200 and empty body if API can passed validate', async () => {

			await handler(ApiDummy, {
				body: {
					name: 'Henry',
					age: 11,
					content: { houseId: 11 }
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 200);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {});
		});

		it('Should return 400 and if API cannot passed validate', async () => {

			await handler(ApiDummy, {
				body: {
					name: 'John',
					age: 11,
					content: { houseId: 11 }
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 400);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Wrong name!'
			});
		});

		it('Should return 400 and if API cannot passed validate with extraMessage', async () => {

			await handler(ApiDummy, {
				body: {
					name: 'Henry',
					age: 10,
					content: { houseId: 11 }
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 400);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Wrong age',
				error: 'Not 10'
			});
		});

		it('Should return 500 and if API cannot passed validate with a typeError', async () => {

			await handler(ApiDummy, {
				body: {
					name: 'Henry',
					age: 11
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 500);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Internal Server Error',
				error: 'Cannot read property \'houseId\' of undefined'
			});
		});

		it('Should return 404 and if API cannot passed validate with custom statusCode', async () => {

			await handler(ApiDummy, {
				body: {
					name: 'Henry',
					age: 11,
					content: { houseId: 10 }
				}
			}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 404);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Not found'
			});
		});
	});

	context('When handler can execute process from API', () => {

		it('Should return 200 and empty body if Class has not struct, or validate, or process', async () => {

			class ApiDummy extends API {}

			await handler(ApiDummy, {}, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 200);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {});
		});

		it('Should return 200 with custom body response', async () => {

			class ApiDummy extends API {
				process() {
					this.setBody({
						method: this.request.method
					});
				}
			}

			await handler(ApiDummy, { method: 'POST' }, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 200);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				method: 'POST'
			});
		});

		it('Should return custom statusCode with custom body response', async () => {

			class ApiDummy extends API {
				process() {
					this.setCode(202).setBody({
						url: this.request.url
					});
				}
			}

			await handler(ApiDummy, { url: '/api/custom' }, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 202);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				url: '/api/custom'
			});
		});

		it('Should return custom headers and do not use .json() if change content-type', async () => {

			class ApiDummy extends API {
				process() {
					this.setHeader('Content-Type', 'text/plain')
						.setHeader('x-custom', 'test')
						.setCode(202);
				}
			}

			await handler(ApiDummy, { url: '/api/custom' }, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 202);
			sinon.assert.calledOnceWithExactly(responseDummy.send, {});
			sinon.assert.notCalled(responseDummy.json);

			sinon.assert.calledTwice(responseDummy.setHeader);
			sinon.assert.calledWithExactly(responseDummy.setHeader, 'Content-Type', 'text/plain');
			sinon.assert.calledWithExactly(responseDummy.setHeader, 'x-custom', 'test');
		});

		it('Should return 500 if process fails', async () => {

			class ApiDummy extends API {
				process() {
					throw new Error('Fail');
				}
			}

			await handler(ApiDummy, { url: '/api/custom' }, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 500);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Fail'
			});
		});

		it('Should rejects with custom statusCode and extraMessage', async () => {

			class ApiDummy extends API {
				process() {
					this.setCode(502);
					const error = new Error('Timeout');
					error.extraMessage = 'Not really';

					throw error;
				}
			}

			await handler(ApiDummy, { url: '/api/custom' }, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 502);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Timeout',
				error: 'Not really'
			});
		});

		it('Should rejects 500 because TypeError', async () => {

			class ApiDummy extends API {
				process() {
					return this.data.content.houseId === 10;
				}
			}

			await handler(ApiDummy, { url: '/api/custom' }, responseDummy);

			sinon.assert.calledOnceWithExactly(responseDummy.status, 500);
			sinon.assert.calledOnceWithExactly(responseDummy.json, {
				message: 'Something went wrong',
				error: 'Cannot read property \'content\' of undefined'
			});
		});
	});
});


tests/parse-query.js
'use strict';

const assert = require('assert');

const parseQuery = require('../lib/helpers/parse-query');

describe('Parse Query', () => {

	context('When Query has Filters', () => {

		it('Should return every query as filters', () => {

			const queries = {
				name: 'John',
				age: '11'
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: queries,
				sort: null,
				sortDirection: null,
				pathIds: {}
			});
		});

		it('Should return every query as filters using filter. notation', () => {

			const queries = {
				'filters.name': 'John',
				'filters.age': '11'
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {
					name: 'John',
					age: '11'
				},
				sort: null,
				sortDirection: null,
				pathIds: {}
			});
		});

		it('Should return every query as filters mixing filter. notation with normal filters', () => {

			const queries = {
				'filters.name': 'John',
				age: '11'
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {
					name: 'John',
					age: '11'
				},
				sort: null,
				sortDirection: null,
				pathIds: {}
			});
		});
	});

	context('When Query has Sort fields', () => {

		it('Should return sort fields with asc direction', () => {

			const queries = {
				sortBy: 'John',
				sortDirection: 'asc'
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {},
				sort: 'John',
				sortDirection: 'asc',
				pathIds: {}
			});
		});

		it('Should return sort fields with asc direction if 1 is passed', () => {

			const queries = {
				sort: 'John',
				sortDirection: 1
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {},
				sort: 'John',
				sortDirection: 'asc',
				pathIds: {}
			});
		});

		it('Should return sort fields with asc direction if not-0 is passed', () => {

			const queries = {
				sortBy: 'John',
				sortDirection: 2
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {},
				sort: 'John',
				sortDirection: 'asc',
				pathIds: {}
			});
		});

		it('Should return sort fields with desc direction if 0 is passed', () => {

			const queries = {
				sort: 'John',
				sortDirection: 0
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {},
				sort: 'John',
				sortDirection: 'desc',
				pathIds: {}
			});
		});

		it('Should return sort fields with desc direction', () => {

			const queries = {
				sortBy: 'John',
				sortDirection: 0
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {},
				sort: 'John',
				sortDirection: 'desc',
				pathIds: {}
			});
		});

		it('Should return the last sort if both form are passed', () => {

			const queries = {
				sortBy: 'John',
				sort: 'Doe'
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {},
				sort: 'Doe',
				sortDirection: null,
				pathIds: {}
			});

			assert.deepStrictEqual(parseQuery({
				sort: 'Doe',
				sortBy: 'John'
			}), {
				filters: {},
				sort: 'John',
				sortDirection: null,
				pathIds: {}
			});
		});
	});

	context('When Query has PathIds', () => {

		it('Should return pathIds when query has pathIds. notation', () => {

			const queries = {
				'pathIds.name': 'John',
				'pathIds.age': '11'
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {},
				sort: null,
				sortDirection: null,
				pathIds: {
					name: 'John',
					age: '11'
				}
			});
		});

		it('Should not return pathIds', () => {

			const queries = {
				'pathId.name': 'John',
				'pathId.age': '11'
			};

			assert.deepStrictEqual(parseQuery(queries), {
				filters: {},
				sort: null,
				sortDirection: null,
				pathIds: {},
				pathId: {
					name: 'John',
					age: '11'
				}
			});
		});
	});
});


lib/api.js
'use strict';

const parseQuery = require('./helpers/parse-query');

/**
 * The Request Data
 * @typedef {Object} Request
 * @property {string} url The relative endpoint
 * @property {string} method The REST Method
 * @property {string} cookies The Cookies
 * @property {string} headers The Headers
 */

/**
 * The Query-Parameters used for Sorting
 * @typedef {Object} Sort
 * @property {string} by The name of the field to Sort By
 * @property {string} direction The direction to sort 'asc' or 'desc'
 */

/**
 * Class to Handle API throw a Serverless Function.
 * @typedef {Class} API
 * @property {Object} data The Request Body Data
 * @property {Object} pathIds The Path-Parameters Ids
 * @property {Object} filters The Query-Parameters used for Filters
 * @property {Object} otherQuery Other Query-Parameters
 * @property {Sort} sort The Query-Parameters used for Sorting
 * @property {Request} request
 */
module.exports = class API {

	constructor({
		url, method, body, query = {}, cookies, headers
	}) {

		const {
			pathIds, filters, sort, sortDirection, ...otherQuery
		} = parseQuery(query);

		this.data = body;
		this.pathIds = pathIds;
		this.filters = filters;
		this.query = otherQuery;

		if(sort) {
			this.sort = {
				by: sort,
				...sortDirection && { direction: sortDirection }
			};
		}

		this.request = {
			url,
			method,
			cookies,
			headers
		};
	}

	/**
	 * Set response status-code
	 * @param {number} code The Status Code
	 * @returns {API}
	 */
	setCode(code) {
		this.statusCode = code;
		return this;
	}

	/**
	 * Set Response Body
	 * @param {*} body The Body to response, if you don't set content-type header must be an Object
	 * @returns {InstAPIanceType}
	 */
	setBody(body) {
		this.responseBody = body;
		return this;
	}

	/**
	 * Set Response Headers
	 * @param {string} header The header you want to set
	 * @param {string|number|boolean} value The value of the header
	 * @returns {API}
	 */
	setHeader(header, value) {

		if(!this.responseHeaders)
			this.responseHeaders = {};

		this.responseHeaders[header] = value;
		return this;
	}
};


lib/handler.js
'use strict';

const Dispatcher = require('./helpers/dispatcher');

/**
 * For Handling a Serverless Function
 * @param {API} ApiClass Class to handler the API
 * @param {Object} request Request Data - Vercel arguments
 * @param {Object} response Response Data - Vercel arguments
 */
module.exports = async (ApiClass, request, response) => {

	if(!response)
		throw new Error('Response is not passed');

	if(!ApiClass) {
		response
			.status(500)
			.json({
				message: 'Internal Server Error',
				error: 'API cannot be initialize. Cannot found API Class'
			});

		return;
	}

	const apiInstance = new ApiClass(request);

	const dispatcher = new Dispatcher(apiInstance, ApiClass, response);

	if(!dispatcher.validateStruct('Ids', apiInstance.pathIds))
		return;

	if(!dispatcher.validateStruct('Body', apiInstance.data))
		return;

	if(!dispatcher.validateStruct('Filters', apiInstance.filters))
		return;

	if(!dispatcher.validateStruct('Sort', apiInstance.sort))
		return;

	if(!dispatcher.validateStruct('Query', apiInstance.query))
		return;

	if(apiInstance.validate && !await dispatcher.validateApi())
		return;

	if(apiInstance.process)
		return dispatcher.processApi();

	response.status(200).json({});
};


lib/helpers/dispatcher.js
'use strict';

module.exports = class Dispatcher {

	constructor(apiInstance, ApiClass, response) {
		this.apiInstance = apiInstance;
		this.ApiClass = ApiClass;
		this.response = response;
	}

	validateStruct(structType, data) {

		const structName = `${structType.toLowerCase()}Struct`;

		if(!this.ApiClass[structName])
			return true;

		const [structError] = this.ApiClass[structName].validate(data);

		if(structError) {
			this.response
				.status(400)
				.json({
					message: `Invalid Struct for ${structType}`,
					error: this.parseStructError(structError)
				});
		}

		return !structError;
	}

	parseStructError({ errors }) {
		/* istanbul ignore next */
		return errors.map(({ type, value, path, reason }) => ({
			field: path.join('.'),
			type,
			value,
			...reason && { reason }
		}));
	}

	async validateApi() {

		try {
			await this.apiInstance.validate();
			return true;

		} catch(error) {

			if(this.isTypeError(error))
				this.apiInstance.statusCode = 500;

			this.response
				.status(this.apiInstance.statusCode || 400)
				.json({
					message: this.isTypeError(error) ? 'Internal Server Error' : error.message,
					...(this.isTypeError(error) || error.extraMessage) && { error: error.extraMessage || error.message }
				});

			return false;
		}
	}

	async processApi() {

		try {

			await this.apiInstance.process();

			// eslint-disable-next-line newline-per-chained-call
			this.setHeaders().status(this.apiInstance.statusCode || 200)[this.selectResponseType()](this.apiInstance.responseBody || {});

		} catch(error) {
			this.response
				.status(this.apiInstance.statusCode || 500)
				.json({
					message: this.isTypeError(error) ? 'Something went wrong' : error.message,
					...(this.isTypeError(error) || error.extraMessage) && { error: error.extraMessage || error.message }
				});
		}
	}

	isTypeError(error) {
		return ['TypeError', 'ReferenceError'].includes(error.name);
	}

	setHeaders() {

		if(this.apiInstance.responseHeaders && Object.keys(this.apiInstance.responseHeaders).length) {
			Object.entries(this.apiInstance.responseHeaders).forEach(([header, value]) => {
				this.response.setHeader(header, value);
			});
		}

		return this.response;
	}

	selectResponseType() {

		if(this.apiInstance.responseHeaders && this.apiInstance.responseHeaders['Content-Type'] !== 'application/json')
			return 'send';

		return 'json';
	}
};
